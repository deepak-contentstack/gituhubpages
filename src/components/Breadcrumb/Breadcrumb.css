@import '../../../tokens/variables.scss';
@import '../../../tokens/tokens.css';
@import '../../static/styles/css/typography.css';
@import '../../static/styles/css/common.css';

.Breadcrumb {
    color: $color-font-base;
    font-size: $size-font-small;
    font-style: normal;
    font-weight: $font-weight-medium;
    letter-spacing: -6e-5em;
    display:flex;
    align-items: center;
    div {
        display: inline;
    }
    &__link{
        display:inline-block;
        color: $color-font-base;
        cursor: pointer;
        font-weight: $font-weight-regular;
        height: $space-15 ;
        
        &--text{
            display:inline-block;
            height: $space-15;
            line-height: $space-15;
            @include text-wrapper;
        }
        &--slash{
            height:$space-15;
            display: inline-block;
        }
    }
    a {
        color: $color-font-tertiary;
        text-decoration: none;
    }
}