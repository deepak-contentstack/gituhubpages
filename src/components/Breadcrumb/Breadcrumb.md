# Breadcrumb

Breadcrumbs are a great way to show hierarchy-based information for a page

## When to use

When you want to indicate the current page's location within a navigational hierarchy.

### Properties

- **breadcrumbItems** - It's an array of objects consisting of path(url) and corresponding display name. Display name can be a string or a react node | component.
  **_*Syntax*_**: breadcrumbItems=[
  {
  name:'Breadcrumb 1'.
  path:'url',
  action:function
  },
  ...
  ]

#### Eg: &lt; Breadcrumb breadcrumbItems={breadcrumbItems} &gt;
