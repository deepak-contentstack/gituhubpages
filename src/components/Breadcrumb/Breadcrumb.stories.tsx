import React from 'react'
import { withA11y } from '@storybook/addon-a11y'
import { withKnobs } from '@storybook/addon-knobs'
import { withRouterBreadcrumb as Breadcrumb } from '../../index'
import notes from './Breadcrumb.md'
import { BrowserRouter as Router } from 'react-router-dom'

export default {
  title: 'In-progress / Breadcrumb',
  component: Breadcrumb,
  decorators: [withKnobs, withA11y],
  parameters: {
    componentSubtitle: 'Breadcrumbs are a great way to show hierarchy-based information for a page',
    notes: { notes },
    design: {
      type: 'figma',
      url: '',
    },
  },
}
const Component = () => <span> Breadcrumb 3</span>

const breadcrumbItems = [
  {
    name: 'Breadcrumb 1',
    path: '#path',
    action: () => console.log('Breadcrumb 1 clicked'),
  },
  {
    name: 'Breadcrumb 2',
    path: '#path1',
    action: () => console.log('Breadcrumb 2 clicked'),
  },
  {
    name: <Component />,
    path: '#path2',
    action: () => console.log('Breadcrumb 3 clicked'),
  },
]

export const Default = () => {
  return (
    <Router>
      <div>
        <Breadcrumb breadcrumbItems={breadcrumbItems} />
      </div>
    </Router>
  )
}

export const WithMaxWidth = () => {
  return (
    <Router>
      <div>
        <Breadcrumb breadcrumbItems={breadcrumbItems} tooltipPosition="top" maxWidth={50}/>
      </div>
    </Router>
  )
}
