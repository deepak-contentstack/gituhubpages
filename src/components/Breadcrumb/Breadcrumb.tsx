import React, { useEffect, useState ,useRef} from 'react'
import { Link, withRouter } from 'react-router-dom'
import './Breadcrumb.css'
import Tooltip  from '../Tooltip/Tooltip'
interface IBreadcrumbItem {
  name: React.ReactNode
  action?: (breadcrumbItem) => void
  path: string
}

interface IProps {
  maxWidth?:number
  tooltipPosition?:string
  testId?:string
  breadcrumbItems: Array<IBreadcrumbItem>
}

const Breadcrumb = (props: IProps) => {
  const { breadcrumbItems = []} = props;
  const totalItemsLength = breadcrumbItems.length;
  const [tooltipFlags, setTooltipFlags] = useState([]);
  const breadcrumbLinkRef: any = useRef([]);

  useEffect(()=>{
    const offsetWidthLimitFlags = [];

    breadcrumbLinkRef?.current?.forEach(function(element: any) {      
      offsetWidthLimitFlags.push(element?.offsetWidth >= props.maxWidth)
    })
    setTooltipFlags([...offsetWidthLimitFlags])
  }, [breadcrumbItems])
  
  return (
    <div className="Breadcrumb" data-test-id={props.testId}>
      {breadcrumbItems.map((breadcrumbItem: IBreadcrumbItem, index: number) => {
        let isCurrentPage = index === totalItemsLength - 1

        if (isCurrentPage) {
          const name: any = breadcrumbItem.name
          return (
            <>
              {tooltipFlags?.length && tooltipFlags[index] ? (
                   <Tooltip content={name} position={props.tooltipPosition}>
                      <span ref={el => breadcrumbLinkRef.current[index] = el} key={index} style={{maxWidth:props.maxWidth}} className="Breadcrumb__link--text">{name}</span>
                   </Tooltip>
                  ) : (<span ref={el => breadcrumbLinkRef.current[index] = el} className="Breadcrumb__link--text" key={index} style={{maxWidth:props.maxWidth}}>{name}</span>)}   
            </>
          )
        } 
        return (
          <>
            <span className="Breadcrumb__link flex-v-center"  key={index} onClick={() => breadcrumbItem.action(breadcrumbItem)}>
              {tooltipFlags?.length && tooltipFlags[index] ? (
                  <Tooltip content={breadcrumbItem.name} position={props.tooltipPosition}>
                  <span ref={el => breadcrumbLinkRef.current[index] = el} className="Breadcrumb__link--text " style={{maxWidth:props.maxWidth}}>{`${breadcrumbItem.name}`}</span>
                  </Tooltip>
                ) : (<span ref={el => breadcrumbLinkRef.current[index] = el} className="Breadcrumb__link--text" style={{maxWidth:props.maxWidth}}>{`${breadcrumbItem.name}`}</span>)}
              <span className="Breadcrumb__link--slash">{`/`}</span>
            </span>
          </>
        )
      })}
    </div>
  )
}

let withRouterBreadcrumb = withRouter(Breadcrumb)
Breadcrumb.defaultProps = {
  maxWidth: 500,
  tooltipPosition: "bottom",
  testId: 'cs-breadcrumb'
}
export default Breadcrumb
export { withRouterBreadcrumb }


// To revert back if anything fails
// // return (
//   //   <div className="Breadcrumb">
//   //     {breadcrumbItems.map((breadcrumbItem: IBreadcrumbItem, index: number) => {
//   //       let isCurrentPage = index === totalItemsLength - 1
//   //       return (
//   //         <span key={index} onClick={breadcrumbItem.action}>
//   //           {!isCurrentPage ? (
//   //             <>
//   //               <Link to={breadcrumbItem.path}> {breadcrumbItem.name}</Link> /
//   //             </>
//   //           ) : (
//   //             <span> {breadcrumbItem.name}</span>
//   //           )}
//   //         </span>
//   //       )
//   //     })}
//   //   </div>
//   // )