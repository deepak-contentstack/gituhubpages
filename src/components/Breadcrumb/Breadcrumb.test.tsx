import React from 'react'
import renderer from 'react-test-renderer'

import Breadcrumb from './Breadcrumb'
import { BrowserRouter as Router } from 'react-router-dom'

const breadcrumbItems = [
  {
    name: 'Breadcrumb 1',
    path: '#',
    action: () => console.log('Breadcrumb 1 clicked'),
  },
  {
    name: 'Breadcrumb 2',
    path: '#',
    action: () => console.log('Breadcrumb 2 clicked'),
  },
  {
    name: 'Breadcrumb 3',
    path: '#',
    action: () => console.log('Breadcrumb 3 clicked'),
  },
]

describe('Breadcrumb', () => {
  it('render default Breadcrumb component', () => {
    const DefaultBreadcrumb = () => {
      return (
        <Router>
          <Breadcrumb breadcrumbItems={breadcrumbItems} />
        </Router>
      )
    }
    const tree = renderer.create(<DefaultBreadcrumb />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
