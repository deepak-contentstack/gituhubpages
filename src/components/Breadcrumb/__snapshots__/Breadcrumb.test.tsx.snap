// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb render default Breadcrumb component 1`] = `
<div
  className="Breadcrumb"
  data-test-id="cs-breadcrumb"
>
  <span
    className="Breadcrumb__link flex-v-center"
    onClick={[Function]}
  >
    <span
      className="Breadcrumb__link--text"
      style={
        Object {
          "maxWidth": 500,
        }
      }
    >
      Breadcrumb 1
    </span>
    <span
      className="Breadcrumb__link--slash"
    >
      /
    </span>
  </span>
  <span
    className="Breadcrumb__link flex-v-center"
    onClick={[Function]}
  >
    <span
      className="Breadcrumb__link--text"
      style={
        Object {
          "maxWidth": 500,
        }
      }
    >
      Breadcrumb 2
    </span>
    <span
      className="Breadcrumb__link--slash"
    >
      /
    </span>
  </span>
  <span
    className="Breadcrumb__link--text"
    style={
      Object {
        "maxWidth": 500,
      }
    }
  >
    Breadcrumb 3
  </span>
</div>
`;
